packages/make-styles-webpack-loader/src/webpackLoader.ts

```diff
@@ -0,0 +1,89 @@
+import { EvalCache, Module } from '@linaria/babel';
+import * as enhancedResolve from 'enhanced-resolve';
+import * as path from 'path';
+import * as webpack from 'webpack';
+
+import { transformSync, TransformResult, TransformOptions } from './transformSync';
+
+type WebpackLoaderParams = Parameters<webpack.LoaderDefinitionFunction<never>>;
+
+/**
+ * Webpack can also pass sourcemaps as a string, Babel accepts only objects.
+ * See https://github.com/babel/babel-loader/pull/889.
+ */
+function parseSourceMap(inputSourceMap: WebpackLoaderParams[1]): TransformOptions['inputSourceMap'] {
+  try {
+    if (typeof inputSourceMap === 'string') {
+      return JSON.parse(inputSourceMap) as TransformOptions['inputSourceMap'];
+    }
+
+    return inputSourceMap as TransformOptions['inputSourceMap'];
+  } catch (err) {
+    return undefined;
+  }
+}
+
+export function webpackLoader(
+  this: webpack.LoaderContext<never>,
+  sourceCode: WebpackLoaderParams[0],
+  inputSourceMap: WebpackLoaderParams[1],
+) {
+  EvalCache.clearForFile(this.resourcePath);
+
+  const resolveOptionsDefaults: webpack.ResolveOptions = {
+    conditionNames: ['require'],
+    extensions: ['.js', '.jsx', '.ts', '.tsx', '.json'],
+  };
+  // ⚠ "this._compilation" limits loaders compatibility, however there seems to be no other way to access Webpack's
+  // resolver.
+  // There is this.resolve(), but it's asynchronous. Another option is to read the webpack.config.js, but it won't work
+  // for programmatic usage. This API is used by many loaders/plugins, so hope we're safe for a while
+  const resolveOptionsFromWebpackConfig: webpack.ResolveOptions = this._compilation?.options.resolve || {};
+
+  const resolveSync = enhancedResolve.create.sync({
+    ...resolveOptionsDefaults,
+    alias: resolveOptionsFromWebpackConfig.alias,
+    modules: resolveOptionsFromWebpackConfig.modules,
+  });
+
+  const originalResolveFilename = Module._resolveFilename;
+
+  let result: TransformResult | null = null;
+  let error: Error | null = null;
+
+  try {
+    // Use Webpack's resolution when evaluating modules in Babel plugin
```

Comment:
why is that needed besides "linaria is using that as well" ? can we add better "docs" ?

Reply:
I updated the comment in 241080d, now it should be clearer.